import pandas as pd
import os
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib.patches as mpatches
import matplotlib.cm as cm
import matplotlib.colors as mcolors
import numpy as np
import re
import argparse
from functools import reduce

def plotNumber(df:pd.DataFrame,outfile:str,controlName:str,experimentName:str):
    """
    plotNumber function: plot the output of CaculateNumber function
    """
    df_melted = df.melt(id_vars=["Sample", "group"], value_vars=["TE", "Gene"], var_name="Category", value_name="Count")
    group_colors = {
    controlName: "blue",
    experimentName: "red"
    }

    plt.figure(figsize=(18, 14),dpi=300)
    sns.set_style("whitegrid")
    ax = sns.barplot(x="Sample", y="Count", hue="Category", data=df_melted, palette="Set2")
    # legend generated by sns
    handles_cat, labels_cat = ax.get_legend_handles_labels()
    # xtick name color
    for xtick, sample in zip(ax.get_xticklabels(), df["Sample"]):
        group = df.loc[df["Sample"] == sample, "group"].values[0]  # 获取 group
        xtick.set_color(group_colors[group])  # 设置颜色
    ax.tick_params(axis='x', labelsize=22)
    ax.tick_params(axis='y', labelsize=22) 
    # legend for xtick
    handles_group = [mpatches.Patch(color=group_colors[controlName], label=controlName),
           mpatches.Patch(color=group_colors[experimentName], label=experimentName)]
    lege_ancor = 1.29
    if experimentName == '2C_GFPp_SA1':
        lege_ancor = 1.39
    legend1 = plt.legend(handles=handles_cat, title="feature", loc="upper right",bbox_to_anchor=(lege_ancor, 0.6),fontsize = 30)  # 原来的 TE / Exon 图例
    legend2 = plt.legend(handles=handles_group, title="Group", loc="lower right",bbox_to_anchor=(lege_ancor, 0.6),fontsize = 30)
    plt.subplots_adjust(right=0.77,bottom=0.19,top=0.9)
    plt.gca().add_artist(legend1)

    plt.xticks(rotation=45, ha="right")
    plt.xlabel("Sample", fontsize=30)
    plt.ylabel("Count", fontsize=30)
    plt.savefig(outfile)

if __name__ == '__main__':
    infiles = ["/ChIP_seq_2/StemCells/RNASNP_PT/output/GSE166216/SNP/NumberBar.csv",
               "/ChIP_seq_2/StemCells/RNASNP_PT/output/GSE183522/SNP/NumberBar.csv",
               "/ChIP_seq_2/StemCells/RNASNP_PT/output/GSE185005/SNP/NumberBar.csv",
               "/ChIP_seq_2/StemCells/RNASNP_PT/output/GSE204801/SNP/NumberBar.csv",
               "/ChIP_seq_2/StemCells/RNASNP_PT/output/GSE224794/SNP/NumberBar.csv"]
    dfAll = []
    groupAnn = {1:['mESC','2C_GFPp_SA1'],2:['EPS','TPS'],3:['mESC','ciTotiSC'],4:['prEpiSC','ci8CLC'],5:['hESC','hTBLC']}
    i = 1
    for infile in infiles:
        sample_id = infile.split("/")[5]
        df = pd.read_csv(infile,sep=",")
        print(sample_id)
        outfile = f"/ChIP_seq_2/StemCells/RNASNP_PT/output/plot/mouse/SNP/total/{sample_id}.png"
        controlName = groupAnn[i][0]
        experimentName = groupAnn[i][1]
        print(controlName,experimentName)
        i += 1
        plotNumber(df,outfile,controlName,experimentName)
        # print(df.head())
        # dfAll.append(df)
    